RecoParticleFlow/PFProducer/src/PFAlgo.cc

```diff
@@ -1200,16 +1251,366 @@ int PFAlgo::decideType(const edm::OwnVector<reco::PFBlockElement>& elements,
+          if (hfemSatellite.first < 0. || caloEnergyTmp < totalChargedMomentum) {
+            LogTrace("PFAlgo|createCandidatesHF")
+                << "\t\t\tactive, adding " << std::get<1>(hfemSatellite.second) << " to HFEM energy, and locking";
+            active[std::get<0>(hfemSatellite.second)] = false;
+            // HfEm is excessive (possible for the first hfemSatellite)
+            if (hfemSatellite.first < 0. && (caloEnergyTmp - totalChargedMomentum) > nsigmaHFEM * TotalError) {
+              // HfEm candidate from excess
+              double energyHfEmExcess = max(caloEnergyTmp - totalChargedMomentum, 0.);
+              double uncalibratedenergyHfEmExcess = max(uncalibratedenergyHfEmTmp - totalChargedMomentum, 0.);
+              unsigned tmpi = reconstructCluster(*eclusterRef, energyHfEmExcess);
+              (*pfCandidates_)[tmpi].setHcalEnergy(0, 0.);
+              (*pfCandidates_)[tmpi].setEcalEnergy(uncalibratedenergyHfEmExcess, energyHfEmExcess);
+              (*pfCandidates_)[tmpi].setHoEnergy(0., 0.);
+              (*pfCandidates_)[tmpi].setPs1Energy(0.);
+              (*pfCandidates_)[tmpi].setPs2Energy(0.);
+              (*pfCandidates_)[tmpi].addElementInBlock(blockref, iHfEm);
+              energyHfEmTmp = max(energyHfEmTmp - energyHfEmExcess, 0.);
+              uncalibratedenergyHfEmTmp = max(uncalibratedenergyHfEmTmp - uncalibratedenergyHfEmExcess, 0.);
+            }
+            energyHfEm = energyHfEmTmp;
+            uncalibratedenergyHfEm = uncalibratedenergyHfEmTmp;
+            energyHfHad = energyHfHadTmp;
+            continue;
+          }
+          break;
+        }  // loop over hfemsattelites ends
+      }    // if HFHAD is excessive or not
+
+      //
+      // Loop over all tracks associated to this HFHAD cluster *again* in order to produce charged hadrons
+      //
+      for (auto const& trk : sortedTracksActive) {
+        unsigned iTrack = trk.second;
+
+        // Sanity check
+        reco::TrackRef trackRef = elements[iTrack].trackRef();
+        assert(!trackRef.isNull());
+
+        //
+        // Reconstructing charged hadrons
+        //
+        unsigned tmpi = reconstructTrack(elements[iTrack]);
+        active[iTrack] = false;
+        (*pfCandidates_)[tmpi].addElementInBlock(blockref, iHfHad);
+        auto myHfEms = associatedHfEms.equal_range(iTrack);
+        for (auto ii = myHfEms.first; ii != myHfEms.second; ++ii) {
+          unsigned iHfEm = ii->second.second;
+          if (active[iHfEm])
+            continue;
+          (*pfCandidates_)[tmpi].addElementInBlock(blockref, iHfEm);
+        }
+        double frac = 0.;
+        if (totalChargedMomentum)
+          frac = trackRef->p() / totalChargedMomentum;
+        (*pfCandidates_)[tmpi].setEcalEnergy(uncalibratedenergyHfEm * frac, energyHfEm * frac);
+        (*pfCandidates_)[tmpi].setHcalEnergy(uncalibratedenergyHfHad * frac, energyHfHad * frac);
+        (*pfCandidates_)[tmpi].setHoEnergy(0., 0.);
```

Comment:
since these three calls (HO, PS energy) happen repeatedly, might be useful to encapsulate them in a helper function (to reduce repetition)

Reply:
Thx. For now, just removed these three lines inside createCandidatesHF, which should just give 0 energies anyway, assuming that's preferred.